bin/oneoff/backfill_data/backfill_script_ids_channel_tokens.rb

```diff
@@ -0,0 +1,211 @@
+# The purpose of this script is to backfill the script_id for channel tokens created earlier.
+#
+# How this script works:
+# This script will iterate over channel tokens and identify the script_ID in several possible ways,
+# 1. If the level only exists in 1 script, use that script_id
+# 2. If an associated user exists, if the user has progress in one script that the level could be
+# associated with, use that script_id
+# 3. If an associated user exists, if the user has one user level that is associated with the channel
+# token level, use the script_id from the user level
+# 4. If an associated user exists but the user has no script progress or user levels possibly associated
+# with the channel token, it's likely that the user visited the level page, which generated a channel token,
+# and then left the page. In this case, assign any of the possible scripts to the channel token
+
+require_relative '../../../dashboard/config/environment'
+require 'cdo/db'
+require 'optparse'
+require 'csv'
+
+# We started writing script_id to the table here: https://github.com/code-dot-org/code-dot-org/pull/39855
+# so we don't need to backfill past that point
+MAX_CHANNEL_TOKEN_ID_FOR_BACKFILL = 303_500_000
+
+# Parse options
+options = {
+  start_id: 1,
+  end_id: MAX_CHANNEL_TOKEN_ID_FOR_BACKFILL,
+  dry_run: false,
+}
+
+OptionParser.new do |opts|
+  opts.banner = <<~BANNER
+    Usage: #{File.basename(__FILE__)} [options]
+
+    This script backfills script_id on the Channel Tokens table.
+
+    Options:
+  BANNER
+
+  opts.on('--start-id=1234567',
+    Integer,
+    'Id of first entry to backfill (inclusive).'
+  ) do |start_id|
+    options[:start_id] = start_id
+  end
+
+  opts.on('--end-id=1234567',
+    Integer,
+    'Id of last entry to backfill (inclusive).'
+  ) do |end_id|
+    options[:end_id] = end_id
+  end
+
+  opts.on('--dry-run',
+    'Enables read-only mode where no changes are written to the database'
+  ) do |dry_run|
+    options[:dry_run] = dry_run
+  end
+
+  opts.on('-h', '--help', 'Prints this help message') do
+    puts opts
+    exit
+  end
+end.parse!
+puts "Options: #{options}"
+options.freeze
+
+# we're storing channel token ids that failed backfill in a csv for easy lookup later
+$csv_filename = "bin/oneoff/backfill_data/channel-token-backfill-failures.csv"
+$start_id = options[:start_id]
+$end_id = options[:end_id]
+$is_dry_run = options[:dry_run]
+
+$unable_to_backfill = 0
+
+$level_to_script_ids = {}
+
+def update_script_ids
+  puts "Backfilling channel token script_ids..."
+  puts "Script started at #{Time.now}"
+
+  CSV.open($csv_filename, "a") do |csv|
+    csv.sync = true
+
+    # find_each uses find_in_batches with a batch size of 1000 (https://apidock.com/rails/ActiveRecord/Batches/find_each)
+    ChannelToken.where(id: $start_id..$end_id).find_each do |channel_token|
+      next if channel_token.script_id.present?
+
+      begin
+        level = channel_token.level
```

Comment:
Does this line result in a query to the database?  I'm wondering if it would be worthwhile to defer this until we need something more than the id from the level and/or to cache level objects.